components/common/src/templating/config.rs

```diff
@@ -588,34 +565,156 @@ fn is_toml_value_a_table(key: &str, table: &toml::value::Table) -> bool {
+/// construct the list of template files
+///
+/// `dir` should be a directory that exists.
+fn load_templates<T>(dir: T) -> Result<TemplateRenderer>
+where
+    T: AsRef<Path>,
+{
+    let mut template = TemplateRenderer::new();
+    if dir.as_ref().is_dir() {
+        load_templates_recurse(&dir, &PathBuf::new(), &mut template)?;
+    } else {
+        debug!(
+            "Configuration directory is not a directory: {:?}",
+            dir.as_ref()
+        );
+    }
+    Ok(template)
+}
+
+/// Helper function for `load_templates` to recurse into the directory structure.
+///
+/// `dir` should be a directory
+fn load_templates_recurse<T>(
+    dir: T,
+    context: &PathBuf,
+    mut template: &mut TemplateRenderer,
+) -> Result<()>
+where
+    T: AsRef<Path>,
+{
+    if let Ok(entries) = std::fs::read_dir(dir) {
+        for entry in entries {
+            if let Ok(entry) = entry {
+                // Skip any entries in the template directory which aren't files or directories
+                match entry.file_type() {
+                    Ok(file_type) => {
+                        if file_type.is_file() {
+                            // We're storing the pathname relative to the input config directory
+                            // as the identifier for the template
+                            let name = entry.file_name().to_string_lossy().into_owned();
+                            let relative_path = context.join(&name);
+
+                            // JW TODO: This error needs improvement. TemplateFileError is too generic.
+                            template
+                                .register_template_file(
+                                    &relative_path.to_string_lossy(),
+                                    &entry.path(),
+                                )
+                                .map_err(Error::TemplateFileError)?;
+                        } else if file_type.is_dir() {
+                            load_templates_recurse(
+                                &entry.path(),
+                                &subdir_path(&context, &entry),
+                                &mut template,
+                            )?;
+                        }
+                    }
+                    Err(e) => {
+                        debug!("Failed to get file metadata for {:?} : {}", entry, e);
+                        continue;
+                    }
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Helper function to construct a new PathBuf for each level of subdir we
+/// iterate into.
+fn subdir_path(context: &PathBuf, entry: &std::fs::DirEntry) -> PathBuf {
+    let mut subdir = context.clone();
+    subdir.push(entry.file_name());
+    subdir
+}
+
+/// Create the appropriate directories between a `root` directory
+/// and a file within that directory structure that we're about
+/// to create.
+///
+fn create_directory_structure<T>(root: T, file: T, user: &str, group: &str) -> Result<()>
```

Comment:
Naming this something like `ensure_directory_structure` might be a little more reflective of the intention of this function.